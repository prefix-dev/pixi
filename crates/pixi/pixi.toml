[package]
name = "pixi"

[package.build.backend]
name = "pixi-build-rust"
version = "*"
