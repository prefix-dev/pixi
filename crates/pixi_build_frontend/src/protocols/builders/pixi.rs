use std::{
    fmt::{self, Display, Formatter},
    path::{Path, PathBuf},
    sync::Arc,
};

use miette::Diagnostic;
use pixi_consts::consts;
use pixi_manifest::{Manifest, PackageManifest, PrioritizedChannel, WorkspaceManifest};
use rattler_conda_types::{ChannelConfig, MatchSpec};
use serde::{de::IntoDeserializer, Deserialize};
use thiserror::Error;
use which::Error;

use crate::{
    jsonrpc::{Receiver, Sender},
    protocols::{InitializeError, JsonRPCBuildProtocol},
    tool::{IsolatedToolSpec, ToolCacheError, ToolSpec},
    BackendOverride, InProcessBackend, ToolContext,
};
// use super::{InitializeError, JsonRPCBuildProtocol};

/// A protocol that uses a pixi manifest to invoke a build backend .
#[derive(Debug)]
pub struct ProtocolBuilder {
    source_dir: PathBuf,
    manifest_path: PathBuf,
    workspace_manifest: WorkspaceManifest,
    package_manifest: PackageManifest,
    override_backend_spec: Option<ToolSpec>,
    channel_config: Option<ChannelConfig>,
    cache_dir: Option<PathBuf>,
}

#[derive(thiserror::Error, Debug, Diagnostic)]
pub enum ProtocolBuildError {
    #[error("failed to setup a build backend, the {} could not be parsed", .0.file_name().and_then(std::ffi::OsStr::to_str).unwrap_or("manifest")
    )]
    #[diagnostic(help("Ensure that the manifest at '{}' is a valid pixi project manifest", .0.display()
    ))]
    FailedToParseManifest(PathBuf, #[diagnostic_source] miette::Report),

    #[error("the {} does not describe a package", .0.file_name().and_then(std::ffi::OsStr::to_str).unwrap_or("manifest")
    )]
    #[diagnostic(help("A [package] section is missing in the manifest"))]
    NotAPackage(PathBuf),
}

#[derive(Debug, Error, Diagnostic)]
pub enum FinishError {
    #[error(transparent)]
    #[diagnostic(transparent)]
    Init(#[from] InitializeError),
    NoBuildSection(PathBuf),
    Tool(ToolCacheError),
    SpecConversionError(pixi_spec::SpecConversionError),
}

impl Display for FinishError {
    fn fmt(&self, f: &mut Formatter<'_>) -> fmt::Result {
        match self {
            FinishError::Init(init) => write!(f, "{init}"),
            FinishError::NoBuildSection(_) => write!(f, "failed to setup a build backend, the project manifest does not contain a [build-system] section"),
            FinishError::Tool(ToolCacheError::Instantiate(tool, err)) => match err {
                Error::CannotGetCurrentDirAndPathListEmpty | Error::CannotFindBinaryPath => write!(f, "failed to setup a build backend, the backend tool '{}' could not be found", tool.display()),
                Error::CannotCanonicalize => write!(f, "failed to setup a build backend, although the backend tool  '{}' can be resolved it could not be canonicalized", tool.display()),
            },
            FinishError::Tool(ToolCacheError::Install(report)) => write!(f, "failed to setup a build backend, the backend tool could not be installed: {}", report),
            FinishError::Tool(ToolCacheError::CacheDir(report)) => write!(f, "failed to setup a build backend, the cache dir could not be discovered: {}", report),
            FinishError::SpecConversionError(err) => write!(f, "failed to setup a build backend, the backend tool spec could not be converted: {}", err),
        }
    }
}

impl ProtocolBuilder {
    /// Constructs a new instance from a manifest.
    pub fn new(
        source_dir: PathBuf,
        manifest_path: PathBuf,
        workspace_manifest: WorkspaceManifest,
        package_manifest: PackageManifest,
    ) -> Self {
        Self {
            source_dir,
            manifest_path,
            workspace_manifest,
            package_manifest,
            override_backend_spec: None,
            channel_config: None,
            cache_dir: None,
        }
    }

    /// Sets an optional backend override.
    pub fn with_backend_override(self, backend_override: Option<BackendOverride>) -> Self {
        Self {
            override_backend_spec: backend_override
                .map(BackendOverride::into_spec)
                .or(self.override_backend_spec),
            ..self
        }
    }

    /// Sets the channel configuration used by this instance.
    pub fn with_channel_config(self, channel_config: ChannelConfig) -> Self {
        Self {
            channel_config: Some(channel_config),
            ..self
        }
    }

    /// Sets the cache directory the backend should use.
    pub fn with_opt_cache_dir(self, cache_dir: Option<PathBuf>) -> Self {
        Self { cache_dir, ..self }
    }

    /// Discovers a pixi project in the given source directory.
    pub fn discover(source_dir: &Path) -> Result<Option<Self>, ProtocolBuildError> {
        if let Some(manifest_path) = find_pixi_manifest(source_dir) {
            match Manifest::from_path(&manifest_path) {
                Ok(manifest) => {
                    // Make sure the manifest describes a package.
                    let Some(package_manifest) = manifest.package else {
                        return Err(ProtocolBuildError::NotAPackage(manifest_path));
                    };

                    let builder = Self::new(
                        source_dir.to_path_buf(),
                        manifest_path,
                        manifest.workspace,
                        package_manifest,
                    );
                    return Ok(Some(builder));
                }
                Err(e) => {
                    return Err(ProtocolBuildError::FailedToParseManifest(
                        manifest_path.to_path_buf(),
                        e,
                    ));
                }
            }
        }
        Ok(None)
    }

    pub async fn finish(
        self,
        tool: Arc<ToolContext>,
        build_id: usize,
    ) -> Result<JsonRPCBuildProtocol, FinishError> {
        let channel_config = self.channel_config.unwrap_or_else(|| {
            ChannelConfig::default_with_root_dir(
                self.manifest_path
                    .parent()
                    .expect("a manifest must always reside in a directory")
                    .to_path_buf(),
            )
        });

        let tool_spec = if let Some(backend_spec) = self.override_backend_spec {
            backend_spec
        } else {
            let build_system = &self.package_manifest.build_system;
            let specs = [(
                &build_system.build_backend.name,
                &build_system.build_backend.spec,
            )]
            .into_iter()
            .chain(build_system.additional_dependencies.iter())
            .map(|(name, spec)| {
                spec.clone()
                    .try_into_nameless_match_spec(&channel_config)
                    .map(|spec| MatchSpec::from_nameless(spec, Some(name.clone())))
            })
            .collect::<Result<_, _>>()
            .map_err(FinishError::SpecConversionError)?;

            // Figure out the channels to use
            let channels = build_system.channels.clone().unwrap_or_else(|| {
                PrioritizedChannel::sort_channels_by_priority(
                    self.workspace_manifest.workspace.channels.iter(),
                )
                .cloned()
                .collect()
            });

            ToolSpec::Isolated(IsolatedToolSpec {
                specs,
                command: build_system.build_backend.name.as_source().to_string(),
                channels,
            })
        };

        let tool = tool
            .instantiate(tool_spec, &channel_config)
            .await
            .map_err(FinishError::Tool)?;

        let configuration = self
            .package_manifest
            .build_system
            .build_backend
            .additional_args
            .map_or(serde_json::Value::Null, |value| {
                let deserializer = value.into_deserializer();
                serde_json::Value::deserialize(deserializer).unwrap_or(serde_json::Value::Null)
            });

        Ok(JsonRPCBuildProtocol::setup(
            self.source_dir,
            self.manifest_path,
            configuration,
            build_id,
            self.cache_dir,
            tool,
        )
        .await?)
    }

    /// Finish the construction of the protocol with the given
    /// [`InProcessBackend`] representing the build backend.
    pub async fn finish_with_ipc(
        self,
        ipc: InProcessBackend,
        build_id: usize,
    ) -> Result<JsonRPCBuildProtocol, FinishError> {
        let configuration = self
            .package_manifest
            .build_system
            .build_backend
            .additional_args
            .map_or(serde_json::Value::Null, |value| {
                let deserializer = value.into_deserializer();
                serde_json::Value::deserialize(deserializer).unwrap_or(serde_json::Value::Null)
            });

        Ok(JsonRPCBuildProtocol::setup_with_transport(
            "<IPC>".to_string(),
            self.source_dir,
            self.manifest_path,
            configuration,
            build_id,
            self.cache_dir,
            Sender::from(ipc.rpc_out),
            Receiver::from(ipc.rpc_in),
            None,
        )
        .await?)
    }
}

/// Try to find a pixi manifest in the given source directory.
fn find_pixi_manifest(source_dir: &Path) -> Option<PathBuf> {
    let pixi_manifest_path = source_dir.join(consts::PROJECT_MANIFEST);
    if pixi_manifest_path.exists() {
        return Some(pixi_manifest_path);
    }

    let pyproject_manifest_path = source_dir.join(consts::PYPROJECT_MANIFEST);
    // TODO: Really check if this is a pixi project.
    if pyproject_manifest_path.is_file() {
        return Some(pyproject_manifest_path);
    }

    None
}

#[cfg(test)]
mod tests {
    use std::path::PathBuf;

    use super::ProtocolBuilder;

    #[test]
    pub fn discover_basic_pixi_manifest() {
        let manifest_dir = PathBuf::from(env!("CARGO_MANIFEST_DIR")).join("tests/basic");
        let manifest_path = super::find_pixi_manifest(&manifest_dir)
            .unwrap_or_else(|| panic!("No manifest found at {}", manifest_dir.display()));
        ProtocolBuilder::discover(&manifest_path).unwrap();
    }
}
