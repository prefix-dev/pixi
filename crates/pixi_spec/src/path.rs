use std::path::Path;

use rattler_conda_types::{package::ArchiveIdentifier, NamelessMatchSpec};
use typed_path::{Utf8NativePathBuf, Utf8TypedPathBuf};

use crate::SpecConversionError;

/// A specification of a package from a git repository.
#[derive(Debug, Clone, Hash, Eq, PartialEq)]
pub struct PathSpec {
    /// The path to the package
    pub path: Utf8TypedPathBuf,
}

impl PathSpec {
    /// Converts this instance into a [`NamelessMatchSpec`] if the path points
    /// to binary archive.
    pub fn try_into_nameless_match_spec(
        self,
        root_dir: &Path,
    ) -> Result<Option<NamelessMatchSpec>, SpecConversionError> {
        if !self.is_binary() {
            // Not a binary package
            return Ok(None);
        }

        // Convert the path to an absolute path based on the root_dir
        let path = if self.path.is_absolute() {
            self.path
        } else if let Ok(user_path) = self.path.strip_prefix("~/") {
            let home_dir = dirs::home_dir()
                .ok_or_else(|| SpecConversionError::InvalidPath(self.path.to_string()))?;
            let Some(home_dir_str) = home_dir.to_str() else {
                return Err(SpecConversionError::NotUtf8RootDir(home_dir));
            };
            Utf8TypedPathBuf::from(home_dir_str)
                .join(user_path)
                .normalize()
        } else {
            let Some(root_dir_str) = root_dir.to_str() else {
                return Err(SpecConversionError::NotUtf8RootDir(root_dir.to_path_buf()));
            };
            let native_root_dir = Utf8NativePathBuf::from(root_dir_str);
            if !native_root_dir.is_absolute() {
                return Err(SpecConversionError::NonAbsoluteRootDir(
                    root_dir.to_path_buf(),
                ));
            }

            native_root_dir.to_typed_path().join(self.path).normalize()
        };

        // Convert the absolute url to a file:// url
        let local_file_url =
            file_url::file_path_to_url(path.to_path()).expect("failed to convert path to file url");

        Ok(Some(NamelessMatchSpec {
            url: Some(local_file_url),
            ..NamelessMatchSpec::default()
        }))
    }

    /// Converts this instance into a [`PathSourceSpec`] if the path points to a
    /// source package. Otherwise, returns this instance unmodified.
    #[allow(clippy::result_large_err)]
    pub fn try_into_source_path(self) -> Result<PathSourceSpec, Self> {
        if self.is_binary() {
            Err(self)
        } else {
            Ok(PathSourceSpec { path: self.path })
        }
    }

    /// Returns true if this path points to a binary archive.
    pub fn is_binary(&self) -> bool {
        self.path
            .file_name()
            .and_then(ArchiveIdentifier::try_from_path)
            .is_some()
    }
}

/// Path to a source package. Different from [`PathSpec`] in that this type only
/// refers to source packages.
#[derive(Debug, Clone, Hash, Eq, PartialEq)]
pub struct PathSourceSpec {
    /// The path to the package. Either a directory or an archive.
    pub path: Utf8TypedPathBuf,
}

impl From<PathSourceSpec> for PathSpec {
    fn from(value: PathSourceSpec) -> Self {
        Self { path: value.path }
    }
}
