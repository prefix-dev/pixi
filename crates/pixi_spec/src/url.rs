use rattler_conda_types::{package::ArchiveIdentifier, NamelessMatchSpec};
use rattler_digest::{Md5Hash, Sha256Hash};
use serde_with::serde_as;
use url::Url;

/// A specification of a package from a URL. This is used to represent both
/// source and binary packages.
#[serde_as]
#[derive(Debug, Clone, Hash, Eq, PartialEq, ::serde::Serialize, ::serde::Deserialize)]
pub struct UrlSpec {
    /// The URL of the package
    pub url: Url,

    /// The md5 hash of the package
    #[serde(skip_serializing_if = "Option::is_none")]
    #[serde_as(as = "Option<rattler_digest::serde::SerializableHash::<rattler_digest::Md5>>")]
    pub md5: Option<Md5Hash>,

    /// The sha256 hash of the package
    #[serde(skip_serializing_if = "Option::is_none")]
    #[serde_as(as = "Option<rattler_digest::serde::SerializableHash::<rattler_digest::Sha256>>")]
    pub sha256: Option<Sha256Hash>,
}

impl UrlSpec {
    /// Converts this instance into a [`NamelessMatchSpec`] if the URL points to
    /// a binary package.
    #[allow(clippy::result_large_err)]
    pub fn try_into_nameless_match_spec(self) -> Result<NamelessMatchSpec, Self> {
        if self.is_binary() {
            Ok(NamelessMatchSpec {
                url: Some(self.url),
                md5: self.md5,
                sha256: self.sha256,
                ..NamelessMatchSpec::default()
            })
        } else {
            Err(self)
        }
    }

    /// Converts this instance into a [`UrlSourceSpec`] if the URL points to a
    /// source package. Otherwise, returns this instance unmodified.
    #[allow(clippy::result_large_err)]
    pub fn try_into_source_url(self) -> Result<UrlSourceSpec, Self> {
        if self.is_binary() {
            Err(self)
        } else {
            Ok(UrlSourceSpec {
                url: self.url,
                md5: self.md5,
                sha256: self.sha256,
            })
        }
    }

    /// Returns true if the URL points to a binary package.
    pub fn is_binary(&self) -> bool {
        ArchiveIdentifier::try_from_url(&self.url).is_some()
    }
}

/// A specification of a source archive from a URL.
#[derive(Debug, Clone, Hash, Eq, PartialEq)]
pub struct UrlSourceSpec {
    /// The URL of the package
    pub url: Url,

    /// The md5 hash of the archive
    pub md5: Option<Md5Hash>,

    /// The sha256 hash of the archive
    pub sha256: Option<Sha256Hash>,
}

impl From<UrlSourceSpec> for UrlSpec {
    fn from(value: UrlSourceSpec) -> Self {
        Self {
            url: value.url,
            md5: value.md5,
            sha256: value.sha256,
        }
    }
}
