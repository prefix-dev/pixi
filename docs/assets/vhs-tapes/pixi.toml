[workspace]
channels = ["https://prefix.dev/conda-forge"]
platforms = ["osx-arm64", "linux-64", "win-64"]

[tasks]
_global-dark = "vhs demo_global_dark.tape"
_global-light = "vhs demo_global_light.tape"
_project-dark = "vhs demo_project_dark.tape"
_project-light = "vhs demo_project_light.tape"
record-all = { depends-on = [
  "_global-light",
  "_global-dark",
  "_project-light",
  "_project-dark",
] }

[dependencies]
ffmpeg = ">=7.1.1,<8"
vhs = ">=0.7.2,<0.8"
[target.unix.dependencies]
ttyd = ">=1.7.7,<2"
