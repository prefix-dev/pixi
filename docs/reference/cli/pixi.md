<!--- This file is autogenerated. Do not edit manually! -->
# <code>pixi</code>

--8<-- "docs/reference/cli/pixi_extender.md:description"

## Usage
```
pixi [OPTIONS] <COMMAND>
```

## Subcommands
| Command | Description |
|---------|-------------|
| [`init`](pixi/init.md) | Creates a new workspace |
| [`add`](pixi/add.md) | Adds dependencies to the workspace |
| [`remove`](pixi/remove.md) | Removes dependencies from the workspace |
| [`install`](pixi/install.md) | Install an environment, both updating the lockfile and installing the environment |
| [`update`](pixi/update.md) | The `update` command checks if there are newer versions of the dependencies and updates the `pixi.lock` file and environments accordingly |
| [`upgrade`](pixi/upgrade.md) | Checks if there are newer versions of the dependencies and upgrades them in the lockfile and manifest file |
| [`lock`](pixi/lock.md) | Solve environment and update the lock file without installing the environments |
| [`run`](pixi/run.md) | Runs task in the pixi environment |
| [`exec`](pixi/exec.md) | Run a command and install it in a temporary environment |
| [`shell`](pixi/shell.md) | Start a shell in a pixi environment, run `exit` to leave the shell |
| [`shell-hook`](pixi/shell-hook.md) | Print the pixi environment activation script |
| [`workspace`](pixi/workspace.md) | Modify the workspace configuration file through the command line |
| [`task`](pixi/task.md) | Interact with tasks in the workspace |
| [`list`](pixi/list.md) | List workspace's packages |
| [`tree`](pixi/tree.md) | Show a tree of workspace dependencies |
| [`global`](pixi/global.md) | Subcommand for global package management actions |
| [`auth`](pixi/auth.md) | Login to prefix.dev or anaconda.org servers to access private channels |
| [`config`](pixi/config.md) | Configuration management |
| [`info`](pixi/info.md) | Information about the system, workspace and environments for the current machine |
| [`upload`](pixi/upload.md) | Upload a conda package |
| [`search`](pixi/search.md) | Search a conda package |
| [`self-update`](pixi/self-update.md) | Update pixi to the latest version or a specific version |
| [`clean`](pixi/clean.md) | Cleanup the environments |
| [`completion`](pixi/completion.md) | Generates a completion script for a shell |
| [`build`](pixi/build.md) | Workspace configuration |


## Global Options
- <a id="arg---help" href="#arg---help">`--help (-h)`</a>
:  Display help information
- <a id="arg---verbose" href="#arg---verbose">`--verbose (-v)`</a>
:  Increase logging verbosity (-v for warnings, -vv for info, -vvv for debug, -vvvv for trace)
- <a id="arg---quiet" href="#arg---quiet">`--quiet (-q)`</a>
:  Decrease logging verbosity (quiet mode)
- <a id="arg---color" href="#arg---color">`--color <COLOR>`</a>
:  Whether the log needs to be colored
<br>**env**: `PIXI_COLOR`
<br>**default**: `auto`
<br>**options**: `always`, `never`, `auto`
- <a id="arg---no-progress" href="#arg---no-progress">`--no-progress`</a>
:  Hide all progress bars, always turned on if stderr is not a terminal
<br>**env**: `PIXI_NO_PROGRESS`
<br>**default**: `false`

--8<-- "docs/reference/cli/pixi_extender.md:example"
