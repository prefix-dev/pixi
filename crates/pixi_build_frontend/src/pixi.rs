use std::path::{Path, PathBuf};

use crate::tool::{IsolatedToolSpec, Tool, ToolSpec};
use crate::Metadata;
use pixi_consts::consts;
use pixi_manifest::Manifest;
use rattler_conda_types::{MatchSpec, ParseStrictness::Strict};

/// A protocol that uses a pixi manifest to invoke a build backend .
#[derive(Debug)]
pub(crate) struct PixiProtocol {
    _manifest: Manifest,
    backend_spec: ToolSpec,
}

impl PixiProtocol {
    /// Constructs a new instance from a manifest.
    pub fn new(manifest: Manifest) -> Self {
        // TODO: Replace this with something that we read from the manifest.
        let backend_spec =
            IsolatedToolSpec::from_specs(vec![
                MatchSpec::from_str("pixi-build-python", Strict).unwrap()
            ])
            .into();

        Self {
            _manifest: manifest,
            backend_spec,
        }
    }

    /// Discovers a pixi project in the given source directory.
    pub fn discover(source_dir: &Path) -> miette::Result<Option<Self>> {
        if let Some(manifest_path) = find_pixi_manifest(source_dir) {
            let manifest = Manifest::from_path(&manifest_path)?;
            return Ok(Some(Self::new(manifest)));
        }
        Ok(None)
    }

    /// Returns the backend spec of this protocol.
    pub fn backend_tool(&self) -> ToolSpec {
        self.backend_spec.clone()
    }

    /// Extract metadata from the recipe.
    pub fn get_metadata(&self, _backend: &Tool) -> miette::Result<Metadata> {
        todo!("extract metadata from pixi manifest")
    }
}

/// Try to find a pixi manifest in the given source directory.
fn find_pixi_manifest(source_dir: &Path) -> Option<PathBuf> {
    let pixi_manifest_path = source_dir.join(consts::PROJECT_MANIFEST);
    if pixi_manifest_path.exists() {
        return Some(pixi_manifest_path);
    }

    let pyproject_manifest_path = source_dir.join(consts::PYPROJECT_MANIFEST);
    // TODO: Really check if this is a pixi project.
    if pyproject_manifest_path.is_file() {
        return Some(pyproject_manifest_path);
    }

    None
}
