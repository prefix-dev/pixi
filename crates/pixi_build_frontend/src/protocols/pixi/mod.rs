mod protocol;

use std::path::{Path, PathBuf};

use miette::IntoDiagnostic;
use pixi_consts::consts;
use pixi_manifest::Manifest;
pub use protocol::{InitializeError, Protocol};
use rattler_conda_types::ChannelConfig;

use crate::tool::{IsolatedToolSpec, Tool, ToolSpec};

/// A protocol that uses a pixi manifest to invoke a build backend .
#[derive(Debug)]
pub(crate) struct ProtocolBuilder {
    source_dir: PathBuf,
    _manifest: Manifest,
    backend_spec: ToolSpec,
    channel_config: ChannelConfig,
}

#[derive(thiserror::Error, Debug)]
pub enum ProtocolBuildError {
    #[error("No build section found")]
    NoBuildSection,
}

impl ProtocolBuilder {
    /// Constructs a new instance from a manifest.
    pub fn new(source_dir: PathBuf, manifest: Manifest) -> Result<Self, ProtocolBuildError> {
        let backend_spec = manifest
            .build_section()
            .map(IsolatedToolSpec::from_build_section)
            .ok_or(ProtocolBuildError::NoBuildSection)?;

        Ok(Self {
            source_dir,
            _manifest: manifest,
            backend_spec: backend_spec.into(),
            channel_config: ChannelConfig::default_with_root_dir(PathBuf::new()),
        })
    }

    /// Sets the channel configuration used by this instance.
    pub fn with_channel_config(self, channel_config: ChannelConfig) -> Self {
        Self {
            channel_config,
            ..self
        }
    }

    /// Discovers a pixi project in the given source directory.
    pub fn discover(source_dir: &Path) -> miette::Result<Option<Self>> {
        if let Some(manifest_path) = find_pixi_manifest(source_dir) {
            let manifest = Manifest::from_path(manifest_path)?;
            return Ok(Some(
                Self::new(source_dir.to_path_buf(), manifest).into_diagnostic()?,
            ));
        }
        Ok(None)
    }

    /// Returns the backend spec of this protocol.
    pub fn backend_tool(&self) -> ToolSpec {
        self.backend_spec.clone()
    }

    pub async fn finish(self, tool: Tool) -> Result<Protocol, InitializeError> {
        Protocol::setup(
            self.source_dir,
            self._manifest.path,
            self.channel_config,
            tool,
        )
        .await
    }
}

/// Try to find a pixi manifest in the given source directory.
fn find_pixi_manifest(source_dir: &Path) -> Option<PathBuf> {
    let pixi_manifest_path = source_dir.join(consts::PROJECT_MANIFEST);
    if pixi_manifest_path.exists() {
        return Some(pixi_manifest_path);
    }

    let pyproject_manifest_path = source_dir.join(consts::PYPROJECT_MANIFEST);
    // TODO: Really check if this is a pixi project.
    if pyproject_manifest_path.is_file() {
        return Some(pyproject_manifest_path);
    }

    None
}

#[cfg(test)]
mod tests {
    use std::path::PathBuf;

    use super::ProtocolBuilder;

    #[test]
    pub fn discover_basic_pixi_manifest() {
        let manifest_dir = PathBuf::from(env!("CARGO_MANIFEST_DIR")).join("tests/basic");
        let manifest_path = super::find_pixi_manifest(&manifest_dir)
            .unwrap_or_else(|| panic!("No manifest found at {}", manifest_dir.display()));
        ProtocolBuilder::discover(&manifest_path).unwrap();
    }
}
