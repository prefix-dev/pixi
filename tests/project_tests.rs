mod common;

use std::path::PathBuf;

use crate::{common::package_database::PackageDatabase, common::PixiControl};
use insta::assert_debug_snapshot;
use pixi::Project;
use rattler_conda_types::{Channel, ChannelConfig, Platform};
use tempfile::TempDir;
use url::Url;

#[tokio::test]
async fn add_channel() {
    // Create a local package database with no entries and write it to disk. This ensures that we
    // have a valid channel.
    let package_database = PackageDatabase::default();
    let initial_channel_dir = TempDir::new().unwrap();
    package_database
        .write_repodata(initial_channel_dir.path())
        .await
        .unwrap();

    // Run the init command
    let pixi = PixiControl::new().unwrap();
    pixi.init()
        .with_local_channel(initial_channel_dir.path())
        .await
        .unwrap();

    // Create and add another local package directory
    let additional_channel_dir = TempDir::new().unwrap();
    package_database
        .write_repodata(additional_channel_dir.path())
        .await
        .unwrap();
    pixi.project_channel_add()
        .with_local_channel(additional_channel_dir.path())
        .await
        .unwrap();

    // There should be a loadable project manifest in the directory
    let project = pixi.project().unwrap();

    // Our channel should be in the list of channels
    let local_channel = Channel::from_str(
        Url::from_directory_path(additional_channel_dir.path()).unwrap(),
        &ChannelConfig::default(),
    )
    .unwrap();
    assert!(project.channels().contains(&local_channel));
}

#[tokio::test]
async fn parse_project() {
    fn dependency_names(project: &Project, platform: Platform) -> Vec<String> {
        project
            .dependencies(None, Some(platform))
            .iter()
            .map(|dep| dep.0.as_normalized().to_string())
            .collect()
    }

    let pixi_toml = include_str!("./pixi_tomls/many_targets.toml");
    let project = Project::from_str(&PathBuf::from("./many"), pixi_toml).unwrap();
    assert_debug_snapshot!(dependency_names(&project, Platform::Linux64));
    assert_debug_snapshot!(dependency_names(&project, Platform::OsxArm64));
    assert_debug_snapshot!(dependency_names(&project, Platform::Win64));
}

#[tokio::test]
async fn parse_valid_schema_projects() {
    // Test all files in the schema/examples/valid directory
    let schema_dir = PathBuf::from(env!("CARGO_MANIFEST_DIR")).join("schema/examples/valid");
    for entry in std::fs::read_dir(schema_dir).unwrap() {
        let entry = entry.unwrap();
        let path = entry.path();
        if path.extension().map(|ext| ext == "toml").unwrap_or(false) {
            let pixi_toml = std::fs::read_to_string(&path).unwrap();
            let _project = Project::from_str(&PathBuf::from(""), &pixi_toml).unwrap();
        }
    }
}
