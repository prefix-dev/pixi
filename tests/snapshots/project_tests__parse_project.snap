---
source: tests/project_tests.rs
expression: "dependency_names(&project, Platform::Linux64)"
---
[
    "all",
    "linux",
    "all_linux",
    "all_unix",
]
